File: bob-server/app/src/main/java/fr/dmconcept/bob/server/activities/MainIOIOActivity.java
package fr.dmconcept.bob.server.activities;

import android.annotation.TargetApi;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.hardware.usb.UsbAccessory;
import android.hardware.usb.UsbManager;
import android.os.Build;
import android.os.Bundle;
import android.os.SystemClock;
import android.provider.Telephony;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import fr.dmconcept.bob.server.BobServer;
import fr.dmconcept.bob.server.R;
import fr.dmconcept.bob.server.ioio.ProjectLock;
import fr.dmconcept.bob.server.models.Project;
import fr.dmconcept.bob.server.models.ServoConfig;
import fr.dmconcept.bob.server.models.Step;
import ioio.lib.api.DigitalOutput;
import ioio.lib.api.IOIO;
import ioio.lib.api.PwmOutput;
import ioio.lib.api.Sequencer;
import ioio.lib.api.exception.ConnectionLostException;
import ioio.lib.util.BaseIOIOLooper;
import ioio.lib.util.IOIOLooper;
import ioio.lib.util.android.IOIOActivity;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Random;
import java.util.Set;

public class MainIOIOActivity extends IOIOActivity {

    private static final String TAG = "BobServer ------------------- ";

    private TextView textViewLog;

    private BobServer mBobServer;

    private ProjectLock mProject = new ProjectLock();

    public static final int PERIOD  = 20000 /* microseconds */; /* 50hz = 0.02s = 20ms = 20.000us */
    public static final int MIN     = 1000 * 2 /* periods */  ; /* 1000us * 0.5us periods         */
    public static final int MAX     = 2000 * 2 /* periods */  ; /* 2000us * 0.5us periods         */
    public static final int PERCENT = new Float((MAX - MIN) / 100f).intValue();

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);

        setContentView(R.layout.main_ioio_activity);
        textViewLog = (TextView) findViewById(R.id.textViewLog);

        info("onCreate()");

        findViewById(R.id.buttonQuit).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

        mBobServer = new BobServer(new BobServer.BobServerListener() {

            @Override
            public void onCantStart(IOException exception) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        new AlertDialog.Builder(MainIOIOActivity.this)
                                .setTitle("Application already running")
                                .setMessage("You can only run one instance of the application")
                                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        finish();
                                    }
                                })
                                .show();
                    }
                });
            }

            @Override
            public void onPlayRequest(Project project) {

//                info("onPlayRequest");

                synchronized (mProject) {

					if (mProject.getProject() != null && !mProject.getProject().boardConfig.id.equals(project.boardConfig.id)) {

                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                new AlertDialog.Builder(MainIOIOActivity.this)
                                        .setTitle("Reconnect the board")
                                        .setMessage("The project played has a different board configuration, you need to restart the app")
                                        .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                                            @Override
                                            public void onClick(DialogInterface dialog, int which) {
                                                finish();
                                            }
                                        })
                                        .show();
                            }
                        });

                        return;
                    }

                    mProject.setProject(project);
                    info("Project " + mProject.getProject().name + " duration=" + mProject.getDuration() + " ms, slices=" + mProject.getSliceCount());
                    mProject.notifyAll();
                }
            }
        });

    }

    private void info(final String message) {

        final String t = Thread.currentThread().getName();

        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (Thread thread: Thread.getAllStackTraces().keySet()) {
            sb.append(thread.getName());
            sb.append(",");
        }
        sb.append("]");

        Log.i(TAG, "-> " + sb.toString());
        Log.i(TAG, t + " - " + message);

        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                textViewLog.setText(textViewLog.getText() + message + "\n");
            }
        });

    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB_MR1)
    @Override
    protected void onNewIntent(Intent intent) {

        super.onNewIntent(intent);

        info("newIntent");

        Log.i(TAG, "MainActivity.onNewIntent(" + intent + ")");
        Bundle extras = intent.getExtras();

        if (!intent.getAction().equals(UsbManager.ACTION_USB_ACCESSORY_ATTACHED)) {
            Log.i(TAG, "MainActivity.onNewIntent() - Not an ACTION_USB_ACCESSORY_ATTACHED action");
            return;
        }

        UsbAccessory accessory = (UsbAccessory) extras.get("accessory");

        if (!accessory.getModel().equals("IOIO")) {
            Log.i(TAG, "MainActivity.onNewIntent() - " + accessory + " don't match the IOIO model");
            return;
        }

    }

    class SequencerLooper extends BaseIOIOLooper {

        private Sequencer sequencer_;

        final Sequencer.Clock CLK = Sequencer.Clock.CLK_2M; /* 0.5us periods */
        final int INITIAL = 3000; /* 1500 us * (1 / 0.5microseconds) */
        final int SLICE =  PERIOD / 16; /* Slice duration in 16microseconds period */

        private ArrayList<Sequencer.ChannelCuePwmPosition> channelCues = new ArrayList();

        @Override
        protected void setup() throws ConnectionLostException, InterruptedException {

            synchronized(mProject) {
                try {
                    info("SequencerLooper.setup() Waiting for a project");
                    mProject.wait();
                } catch (Exception e) {
                    info("Waiting on the project failed");
                }
            }

            info("SequencerLooper.setup() Setting config " + mProject.getProject().boardConfig.name);

            ArrayList<Sequencer.ChannelConfigPwmPosition> channelConfigs = new ArrayList();

            for (ServoConfig sc: mProject.getProject().boardConfig.servoConfigs) {

                // Create the channel configs
                channelConfigs.add(
                    new Sequencer.ChannelConfigPwmPosition(CLK, PERIOD, INITIAL, new DigitalOutput.Spec(sc.port))
                );

                // Create the cues
                channelCues.add(new Sequencer.ChannelCuePwmPosition());

            }

            /*
            final Sequencer.ChannelConfigBinary      led1Config   = new Sequencer.ChannelConfigBinary(true, true, new DigitalOutput.Spec(11));
            */

            sequencer_ = ioio_.openSequencer(channelConfigs.toArray(new Sequencer.ChannelConfig[channelConfigs.size()]));

            // Pre-fill.
            sequencer_.waitEventType(Sequencer.Event.Type.STOPPED);
            while (sequencer_.available() > 0) {
                push();
            }

            info("SequencerLooper.start()");

            sequencer_.start();
        }

        @Override
        public void loop() throws ConnectionLostException, InterruptedException {
            push();
        }

        private long lastLog = 0;
        private int currentSlice = 0;


        private void push() throws ConnectionLostException, InterruptedException {

            ArrayList<Integer> slices = mProject.getSlices().get(currentSlice);

            for (int i = 0; i < channelCues.size(); i++)
                channelCues.get(i).pulseWidth = slices.get(i);

            sequencer_.push(channelCues.toArray(new Sequencer.ChannelCuePwmPosition[channelCues.size()]), SLICE /* 20ms */); // Unit value is 16microseconds, 62500 = 1 s

            /*
            led1Cue.value = new Random().nextBoolean();
            */

            long now = SystemClock.elapsedRealtime();
            if (now - lastLog > 500) {
                lastLog = now;
                Log.i(TAG, "#" + currentSlice + " " + slices);
            }

            currentSlice++;

            if (currentSlice == mProject.getSliceCount()) {

                info("********** LAST SLICE");

                sequencer_.stop();

                synchronized (mProject) {
                    info("SequencerLooper.push() Waiting for a new project");
                    mProject.wait();

                    info("SequencerLooper.push() Got a new project: " + mProject.getDuration() + "ms, " + mProject.getSliceCount() + " slices");
                    currentSlice = 0;

                    info("SequencerLooper.push() Wait for sequencer to stop...");
                    sequencer_.waitEventType(Sequencer.Event.Type.STOPPED);

                    info("SequencerLooper.push() Prefilling sequencer...");
                    while (sequencer_.available() > 0) {
                        push();
                    }

                    info("SequencerLooper.push() Starting sequencer");
                    sequencer_.start();
                }
            }

        }
    }

    class RandomLooper extends BaseIOIOLooper {

        PwmOutput pin3;
        PwmOutput pin4;
        PwmOutput pin5;
        PwmOutput pin6;
        PwmOutput pin7;
        PwmOutput pin10;
        DigitalOutput pin11;

        @Override
        protected void setup() throws ConnectionLostException, InterruptedException {
            pin3 = ioio_.openPwmOutput(3, 50);
            pin4 = ioio_.openPwmOutput(4, 50);
            pin5 = ioio_.openPwmOutput(5, 50);
            pin6 = ioio_.openPwmOutput(6, 50);
            pin7 = ioio_.openPwmOutput(7, 50);
            pin10 = ioio_.openPwmOutput(10, 50);
            pin11 = ioio_.openDigitalOutput(11);
        }

        private int getRandom() {
            return new Random().nextInt(2000 - 1000) + 1000;
        }

        @Override
        public void loop() throws ConnectionLostException, InterruptedException {

            final int servo1 = getRandom();
            final int servo2 = getRandom();
            final int servo3 = getRandom();
            final int servo4 = getRandom();
            final int servo5 = getRandom();
            final int servo6 = getRandom();
            final boolean led1 = new Random().nextBoolean();

            pin3.setPulseWidth(servo1);
            pin4.setPulseWidth(servo2);
            pin5.setPulseWidth(servo3);
            pin6.setPulseWidth(servo4);
            pin7.setPulseWidth(servo5);
            pin10.setPulseWidth(servo6);
            pin11.write(led1);

            Thread.sleep(1000);
        }

        @Override
        public void disconnected() {
            info("disconnected");
        }

        @Override
        public void incompatible() {
            info("incompatible");
        }
    }

    protected IOIOLooper createIOIOLooper() {

        return new SequencerLooper();

    }

    @Override
    protected void onStart() {
        super.onStart();
        info("onStart");
    }

    @Override
    protected void onStop() {

        info("onStop");
        mBobServer.stop();

        super.onStop();
    }

}

File: bob-server/src/fr/dmconcept/bob/server/activities/MainIOIOActivity.scala
package fr.dmconcept.bob.server.activities

import ioio.lib.util.android.IOIOActivity
import android.widget.TextView
import java.util
import fr.dmconcept.bob.server.{R, BobServer}
import fr.dmconcept.bob.server.ioio.ProjectLock
import android.os.{SystemClock, Bundle}
import android.view.View
import android.util.Log
import MainIOIOActivity._
import android.app.AlertDialog
import android.content.{Intent, DialogInterface}
import java.io.IOException
import java.net.{SocketException, InetAddress, NetworkInterface}
import java.util.Collections
import org.apache.http.conn.util.InetAddressUtils
import android.hardware.usb.{UsbAccessory, UsbManager}
import ioio.lib.util.{BaseIOIOLooper, IOIOLooper}
import ioio.lib.api.{DigitalOutput, Sequencer}
import fr.dmconcept.bob.client.models.{ServoConfig, Project}
import collection.JavaConverters._

object MainIOIOActivity {

  final val TAG = "BobServer ------------------- "

  final val PERIOD  = 20000 /* microseconds */ /* 50hz = 0.02s = 20ms = 20.000us */
  final val MIN     = 1000 * 2 /* periods */   /* 1000us * 0.5us periods         */
  final val MAX     = 2000 * 2 /* periods */   /* 2000us * 0.5us periods         */
  final val PERCENT = ((MAX - MIN).toFloat / 100f).intValue()

}

class MainIOIOActivity extends IOIOActivity {

  var textViewIP  : TextView = null
  var textViewLog : TextView = null

  var ips: util.ArrayList[String] = null
  var mBobServer: BobServer = null

  val mProject = new ProjectLock()

  override def onCreate(savedInstanceState: Bundle) {

    super.onCreate(savedInstanceState)

    setContentView(R.layout.main_ioio_activity)

    textViewIP  = findViewById(R.id.textViewIP).asInstanceOf[TextView]
    textViewLog = findViewById(R.id.textViewLog).asInstanceOf[TextView]

    info("onCreate()")

    findViewById(R.id.buttonQuit).setOnClickListener(new View.OnClickListener() {
      override def onClick(v: View): Unit = finish()
    })

    // Get the device IP addresses
    ips = getIPs()

    // Display an error message and exit the app if the device has no IP address
    if (ips.isEmpty()) {
      runOnUiThread(new Runnable() {

        override def run() {

          new AlertDialog.Builder(MainIOIOActivity.this)
            .setTitle("No network connection")
            .setMessage("The app needs a Wifi or Ethernet network connection")
            .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
              override def onClick(dialog: DialogInterface, wich: Int): Unit = finish()
            })
          .show()
        }
      })

      return;

    }

    /*
    mBobServer = new BobServer(new BobServer.BobServerListener() {

      override def onServerStarted() {

        // Display the IP addresses
        textViewIP.setText("");

        ips.asScala.foreach(ip =>
          textViewIP.setText(s"${textViewIP.getText} $ip\n")
        )

      }

      override def onCantStart(exception: IOException) {

        runOnUiThread(new Runnable() {

          override def run() {
            new AlertDialog.Builder(MainIOIOActivity.this)
              .setTitle("Application already running")
              .setMessage("You can only run one instance of the application")
              .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                override def onClick(dialog: DialogInterface, which: Int): Unit = finish()
              })
              .show()
          }
        })

      }

      override def onPlayRequest(project: Project) {

        info(s"onPlayRequest($project.id}")

        /*
        synchronized (mProject) {

          if (mProject.getProject() != null && !mProject.getProject().boardConfig.id.equals(project.boardConfig.id)) {

            runOnUiThread(new Runnable() {
              override def run {
                new AlertDialog.Builder(MainIOIOActivity.this)
                  .setTitle("Reconnect the board")
                  .setMessage("The project played has a different board configuration, you need to restart the app")
                  .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    override def onClick(dialog: DialogInterface, which: Int): Unit = finish()
                  })
                  .show()
              }
            })

            return
          }

          mProject.setProject(project)
          info(s"Project ${mProject.getProject().name} duration=${mProject.getDuration()} ms, slices=${mProject.getSliceCount()}")
          mProject.notifyAll()
        }
        */
      }

    })
  */

  }

  private def info(message: String) {

    Log.i(TAG, message)

    runOnUiThread(new Runnable() {
      override def run = textViewLog.setText(textViewLog.getText() + message + "\n")
    })

  }

  override def onNewIntent(intent: Intent) {

    super.onNewIntent(intent)

    info("MainActivity.onNewIntent(" + intent + ")")

    val extras = intent.getExtras()

    if (!intent.getAction().equals(UsbManager.ACTION_USB_ACCESSORY_ATTACHED)) {
      Log.i(TAG, "MainActivity.onNewIntent() - Not an ACTION_USB_ACCESSORY_ATTACHED action")
      return
    }

    val usbAccessory = extras.get("accessory").asInstanceOf[UsbAccessory]

    if (!usbAccessory.getModel().equals("IOIO")) {
      info(s"MainActivity.onNewIntent() - $usbAccessory don't match the IOIO model")
      return
    }

    //TODO better newIntent handling
    createIOIOLooper()

  }
override def createIOIOLooper: IOIOLooper = return new SequencerLooper()

  object SequencerLooper {
    final val CLK     = Sequencer.Clock.CLK_2M; /* 0.5us periods */
    final val INITIAL = 3000; /* 1500 us * (1 / 0.5microseconds) */
    final val SLICE   =  PERIOD / 16; /* Slice duration in 16microseconds period */
  }

  private class SequencerLooper extends BaseIOIOLooper {

    import SequencerLooper._

    var sequencer_ : Sequencer = null

    val channelCues = new util.ArrayList[Sequencer.ChannelCuePwmPosition]()

    override def setup() /*throws ConnectionLostException, InterruptedException*/ {

      mProject.synchronized {
        info("SequencerLooper.setup() Waiting for a project")
        mProject.wait()
      }

      info(s"SequencerLooper.setup() Setting config ${mProject.project.boardConfig.name}")

      val channelConfigs = new util.ArrayList[Sequencer.ChannelConfigPwmPosition]()

      for (sc: ServoConfig <- mProject.project.boardConfig.servoConfigs) {

        // Create the channel configs
        channelConfigs.add(
          new Sequencer.ChannelConfigPwmPosition(CLK, PERIOD, INITIAL, new DigitalOutput.Spec(sc.pin))
        )

        // Create the cues
        channelCues.add(new Sequencer.ChannelCuePwmPosition())

      }

      sequencer_ = ioio_.openSequencer(channelConfigs.toArray(new Array[Sequencer.ChannelConfig](channelConfigs.size)))

      // Pre-fill.
      sequencer_.waitEventType(Sequencer.Event.Type.STOPPED)
      while (sequencer_.available() > 0) {
        push()
      }

      info("SequencerLooper.start()")

      sequencer_.start()
    }

    override def loop() /* throws ConnectionLostException, InterruptedException */ {
      push()
    }

    var lastLog      = 0l
    var currentSlice = 0

    private def push() /* throws ConnectionLostException, InterruptedException */ {

      val slices = mProject.slices.get(currentSlice)

      for (i <- 0 to channelCues.size())
        channelCues.get(i).pulseWidth = slices.get(i)

      val c = channelCues.toArray(new Array[Sequencer.ChannelCue](channelCues.size))
      sequencer_.push(c, SLICE /* 20ms */); // Unit value is 16microseconds, 62500 = 1 s

      /*
      led1Cue.value = new Random().nextBoolean();
      */

      val now = SystemClock.elapsedRealtime()

      if (now - lastLog > 500) {
        lastLog = now
        Log.i(TAG, "#" + currentSlice + " " + slices)
      }

      currentSlice = currentSlice + 1

      if (currentSlice == mProject.sliceCount) {

        sequencer_.stop()

        mProject.synchronized {
          info("SequencerLooper.push() Sequence done, waiting for a new project")
          mProject.wait()

          info(s"SequencerLooper.push() Got a new project: ${mProject.duration} ms, ${mProject.sliceCount} slices")
          currentSlice = 0

          info("SequencerLooper.push() Wait for sequencer to stop...")
          sequencer_.waitEventType(Sequencer.Event.Type.STOPPED)

          info("SequencerLooper.push() Prefilling the sequencer...")
          while (sequencer_.available() > 0) {
            push()
          }

          info("SequencerLooper.push() Starting the sequencer")
          sequencer_.start()
        }
      }

    }

    override def disconnected(): Unit = super.disconnected()

    override def incompatible(): Unit = super.incompatible()

  }

  override def onStart() {
    super.onStart()
    info("onStart")
  }

  override def onStop() {
    info("onStop")
    mBobServer.stop()

    super.onStop()
  }

  private def getIPs() : util.ArrayList[String] = {

    val ips = new util.ArrayList[String]()

    try {

      NetworkInterface.getNetworkInterfaces().asScala.foreach { intf =>

        val addrs = Collections.list(intf.getInetAddresses())

        addrs.asScala.foreach { addr =>

          if (!addr.isLoopbackAddress()) {

            val sAddr = addr.getHostAddress().toUpperCase()
            val isIPv4 = InetAddressUtils.isIPv4Address(sAddr)

            if (isIPv4)
              ips.add(sAddr)
          }
        }
      }

    } catch {
      case e: SocketException => {}
        // Do notihgn
    }

    ips

  }

}
